File: apps/api/src/api/schemas/meal.py
"""學餐訂購系統 Pydantic Schemas"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from api.models.meal import MealOrderStatus

# `date` 欄位名稱與 datetime.date 型別同名，在 `from __future__ import annotations`
# 環境下 Pydantic 無法分辨。用別名 _Date 指向型別，避免衝突。
_Date = date

# ── 商家 ─────────────────────────────────────────────────────────────────────


class MealVendorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    contact_phone: str | None
    contact_email: str | None
    is_active: bool
    status: str = "approved"
    review_note: str | None = None
    org_id: uuid.UUID
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class MealVendorCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="商家名稱")
    description: str | None = Field(None, description="商家描述")
    contact_phone: str | None = Field(None, max_length=20, description="聯絡電話")
    contact_email: EmailStr | None = Field(None, description="聯絡信箱")
    org_id: uuid.UUID | None = Field(None, description="所屬組織 ID；未提供時系統自動建立")
    manager_email: EmailStr | None = Field(None, description="建立時同步指派的商家負責人")
    status: str | None = Field(None, description="審核狀態；管理員手動新增預設 approved")


class MealVendorUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    contact_phone: str | None = Field(None, max_length=20)
    contact_email: EmailStr | None = None
    is_active: bool | None = None
    status: str | None = None
    review_note: str | None = None


class MealVendorApplicationCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    contact_name: str | None = Field(None, max_length=100)
    contact_phone: str | None = Field(None, max_length=20)
    contact_email: EmailStr | None = None
    org_id: uuid.UUID | None = None


class MealVendorApplicationReview(BaseModel):
    approved: bool
    review_note: str | None = None


class MealVendorApplicationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    contact_name: str | None
    contact_phone: str | None
    contact_email: str | None
    org_id: uuid.UUID
    status: str
    review_note: str | None
    reviewed_by_id: uuid.UUID | None
    reviewed_at: datetime | None
    vendor_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class MealProductCreate(BaseModel):
    vendor_id: uuid.UUID
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    category: str | None = Field(None, max_length=80)
    image_url: str | None = Field(None, max_length=500)
    price: int = Field(..., ge=0)
    default_max_quantity: int | None = Field(None, ge=1)


class MealProductUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    category: str | None = Field(None, max_length=80)
    image_url: str | None = Field(None, max_length=500)
    price: int | None = Field(None, ge=0)
    default_max_quantity: int | None = Field(None, ge=1)
    is_active: bool | None = None


class MealProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    vendor_id: uuid.UUID
    name: str
    description: str | None
    category: str | None
    image_url: str | None
    price: int
    default_max_quantity: int | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class MealPickupSlotCreate(BaseModel):
    label: str = Field(..., min_length=1, max_length=80)
    sort_order: int = 0
    pickup_start: datetime
    pickup_end: datetime
    order_deadline: datetime
    capacity: int | None = Field(None, ge=1)


class MealPickupSlotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    availability_id: uuid.UUID
    label: str
    sort_order: int
    pickup_start: datetime
    pickup_end: datetime
    order_deadline: datetime
    capacity: int | None
    is_active: bool


class MealAvailabilityCreate(BaseModel):
    product_id: uuid.UUID
    service_date: _Date
    sale_start: datetime | None = None
    sale_end: datetime | None = None
    price: int | None = Field(None, ge=0)
    max_quantity: int | None = Field(None, ge=1)
    note: str | None = None
    pickup_slots: list[MealPickupSlotCreate] = Field(default_factory=list)


class MealWeeklyAvailabilityCreate(BaseModel):
    product_ids: list[uuid.UUID] = Field(..., min_length=1)
    date_from: _Date
    date_to: _Date
    weekdays: list[int] = Field(..., min_length=1, description="0=Monday ... 6=Sunday")
    sale_start_time: str | None = Field(None, description="HH:MM")
    sale_end_time: str | None = Field(None, description="HH:MM")
    pickup_slots: list[MealPickupSlotCreate] = Field(default_factory=list)


class MealAvailabilityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    vendor_id: uuid.UUID
    service_date: _Date
    sale_start: datetime | None
    sale_end: datetime | None
    price: int
    max_quantity: int | None
    is_available: bool
    note: str | None
    product: MealProductOut | None = None
    pickup_slots: list[MealPickupSlotOut] = []


# ── 菜單排程 ─────────────────────────────────────────────────────────────────


class MenuItemSummary(BaseModel):
    """用於巢狀在 MenuScheduleOut 中的精簡品項"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    price: int
    max_quantity: int | None
    is_available: bool


class MenuScheduleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    vendor_id: uuid.UUID
    date: _Date
    order_open_time: datetime | None
    order_deadline: datetime
    is_closed: bool
    note: str | None
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    items: list[MenuItemSummary] = []


class MenuScheduleListItem(BaseModel):
    """列表輕量版（不含品項）"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    vendor_id: uuid.UUID
    date: _Date
    order_open_time: datetime | None
    order_deadline: datetime
    is_closed: bool
    note: str | None
    created_at: datetime


class MenuScheduleCreate(BaseModel):
    vendor_id: uuid.UUID = Field(..., description="商家 ID")
    date: _Date = Field(..., description="服務日期")
    order_open_time: datetime | None = Field(None, description="開放訂餐時間（NULL=立即開放）")
    order_deadline: datetime = Field(..., description="結單截止時間（帶時區）")
    note: str | None = Field(None, max_length=500, description="備註")


class MenuScheduleUpdate(BaseModel):
    order_open_time: datetime | None = None
    order_deadline: datetime | None = None
    note: str | None = None


# ── 菜單品項 ─────────────────────────────────────────────────────────────────


class MenuItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    schedule_id: uuid.UUID | None
    name: str
    description: str | None
    price: int
    max_quantity: int | None
    is_available: bool
    created_at: datetime
    updated_at: datetime


class MenuItemCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="品項名稱")
    description: str | None = Field(None, description="品項描述")
    price: int = Field(..., ge=0, description="售價（新台幣）")
    max_quantity: int | None = Field(None, ge=1, description="最大數量（留空代表無限量）")


class MenuItemUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    price: int | None = Field(None, ge=0)
    max_quantity: int | None = Field(None, ge=1)
    is_available: bool | None = None


# ── 學餐訂單 ─────────────────────────────────────────────────────────────────


class MealOrderItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    menu_item_id: uuid.UUID | None = None
    availability_id: uuid.UUID | None = None
    product_name_snapshot: str | None = None
    quantity: int
    unit_price: int
    subtotal: int = 0

    @classmethod
    def model_validate(cls, obj, **kwargs):  # type: ignore[override]
        instance = super().model_validate(obj, **kwargs)
        instance.subtotal = instance.quantity * instance.unit_price
        return instance


class MealOrderItemCreate(BaseModel):
    menu_item_id: uuid.UUID | None = Field(None, description="舊版菜單品項 ID")
    availability_id: uuid.UUID | None = Field(None, description="商品上架 ID")
    quantity: int = Field(..., ge=1, le=20, description="購買數量")


class MealOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    serial_number: str
    pickup_code: str
    user_id: uuid.UUID
    schedule_id: uuid.UUID | None
    vendor_id: uuid.UUID
    availability_id: uuid.UUID | None = None
    pickup_slot_id: uuid.UUID | None = None
    class_id: uuid.UUID | None = None
    status: MealOrderStatus
    total_price: int
    is_paid: bool = False
    paid_at: datetime | None = None
    pickup_status: str = "not_picked"
    pickup_at: datetime | None = None
    notes: str | None
    reminder_sent_at: datetime | None = None
    is_no_show: bool = False
    created_at: datetime
    updated_at: datetime
    items: list[MealOrderItemOut] = []


class MealOrderListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    serial_number: str
    pickup_code: str
    user_id: uuid.UUID
    schedule_id: uuid.UUID
    vendor_id: uuid.UUID
    pickup_slot_id: uuid.UUID | None = None
    class_id: uuid.UUID | None = None
    status: MealOrderStatus
    total_price: int
    is_paid: bool = False
    pickup_status: str = "not_picked"
    is_no_show: bool = False
    created_at: datetime


class MealOrderCreate(BaseModel):
    schedule_id: uuid.UUID | None = Field(None, description="舊版菜單排程 ID")
    pickup_slot_id: uuid.UUID | None = Field(None, description="取餐時段 ID")
    items: list[MealOrderItemCreate] = Field(..., min_length=1, description="訂單品項（至少一項）")
    notes: str | None = Field(None, max_length=500, description="備註")


class MealOrderCancelRequest(BaseModel):
    reason: str | None = Field(None, max_length=500, description="取消原因")


# ── 熱門排序 / 核銷 ───────────────────────────────────────────────────────────


class ItemStatOut(BaseModel):
    """排程品項訂購統計（供前端熱門排序）"""

    item_id: uuid.UUID
    item_name: str
    total_ordered: int


class PickupListItemOut(BaseModel):
    """排程領餐名單列表項"""

    order_id: uuid.UUID
    serial_number: str
    pickup_code: str
    status: MealOrderStatus
    total_price: int
    notes: str | None
    created_at: datetime
    display_name: str
    student_id: str | None
    is_no_show: bool


class MealClassPickupCodeOut(BaseModel):
    code: str
    class_id: uuid.UUID
    vendor_id: uuid.UUID
    pickup_slot_id: uuid.UUID
    expires_at: datetime | None
    order_count: int


class MealPickupLookupOut(BaseModel):
    kind: str
    code: str
    matched_orders: int
    completed_orders: int
    total_price: int
    message: str


class VendorManagerAssignRequest(BaseModel):
    email: str = Field(..., description="要指派為學餐管理員的使用者 Email")


class VendorManagerOut(BaseModel):
    user_id: uuid.UUID
    display_name: str
    email: str
    position_id: uuid.UUID
    user_position_id: uuid.UUID

File: apps/web/src/lib/api/meal.ts
import type {
  ItemStatOut, MealAvailabilityOut, MealClassPickupCodeOut, MealOrderListItem, MealOrderOut, MealPickupLookupOut, MealProductOut, MealVendorApplicationOut, MealVendorOut, MenuItemOut, MenuScheduleListItem, MenuScheduleOut, PickupListItemOut, VendorManagerOut,
} from "../types";
import { BASE, get, post, patch, del, csrfHeaders, silentRefresh, errorMessageFromResponse, ApiError } from "./core";

// ── 學餐系統 ──────────────────────────────────────────────────────────────────

export const mealApi = {
  // 商家
  listVendors: (params?: { org_id?: string; active_only?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.org_id) q.set("org_id", params.org_id);
    if (params?.active_only !== undefined) q.set("active_only", String(params.active_only));
    return get<MealVendorOut[]>(`/meal/vendors?${q}`);
  },

  // 菜單排程
  listSchedules: (params?: { vendor_id?: string; is_closed?: boolean; date_from?: string; date_to?: string }) => {
    const q = new URLSearchParams();
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.is_closed !== undefined) q.set("is_closed", String(params.is_closed));
    if (params?.date_from) q.set("date_from", params.date_from);
    if (params?.date_to) q.set("date_to", params.date_to);
    return get<MenuScheduleListItem[]>(`/meal/schedules?${q}`);
  },
  getSchedule: (id: string) => get<MenuScheduleOut>(`/meal/schedules/${id}`),
  createSchedule: (body: {
    vendor_id: string; date: string;
    order_open_time?: string | null; order_deadline: string; note?: string;
  }) => post<MenuScheduleOut>("/meal/schedules", body),
  updateSchedule: (id: string, body: {
    order_open_time?: string | null; order_deadline?: string; note?: string | null;
  }) => patch<MenuScheduleOut>(`/meal/schedules/${id}`, body),
  closeSchedule: (id: string) => post<MenuScheduleOut>(`/meal/schedules/${id}/close`),
  addMenuItem: (scheduleId: string, body: {
    name: string; description?: string; price: number; max_quantity?: number | null;
  }) => post<MenuItemOut>(`/meal/schedules/${scheduleId}/items`, body),
  updateMenuItem: (itemId: string, body: {
    name?: string; description?: string | null; price?: number;
    max_quantity?: number | null; is_available?: boolean;
  }) => patch<MenuItemOut>(`/meal/items/${itemId}`, body),
  deleteMenuItem: (itemId: string) => del<void>(`/meal/items/${itemId}`),

  // 商家管理
  createVendor: (body: {
    name: string; org_id?: string | null; description?: string | null;
    contact_phone?: string | null; contact_email?: string | null; manager_email?: string | null;
    status?: string | null;
  }) =>
    post<MealVendorOut>("/meal/vendors", body),
  updateVendor: (id: string, body: {
    name?: string; description?: string | null;
    contact_phone?: string | null; contact_email?: string | null; is_active?: boolean;
    status?: string; review_note?: string | null;
  }) => patch<MealVendorOut>(`/meal/vendors/${id}`, body),
  createVendorApplication: (body: {
    name: string; org_id?: string | null; description?: string | null;
    contact_name?: string | null; contact_phone?: string | null; contact_email?: string | null;
  }) => post<MealVendorApplicationOut>("/meal/vendor-applications", body),
  listVendorApplications: (params?: { status?: string; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.status) q.set("status", params.status);
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    if (params?.offset !== undefined) q.set("offset", String(params.offset));
    return get<MealVendorApplicationOut[]>(`/meal/vendor-applications?${q}`);
  },
  reviewVendorApplication: (id: string, body: { approved: boolean; review_note?: string | null }) =>
    post<MealVendorApplicationOut>(`/meal/vendor-applications/${id}/review`, body),
  listVendorManagers: (vendorId: string) =>
    get<VendorManagerOut[]>(`/meal/vendors/${vendorId}/managers`),
  removeVendorManager: (vendorId: string, userId: string) =>
    del<void>(`/meal/vendors/${vendorId}/managers/${userId}`),
  uploadImage: async (file: File): Promise<{ url: string }> => {
    const fd = new FormData();
    fd.append("file", file);
    const doFetch = () =>
      fetch(`${BASE}/meal/images`, {
        method: "POST",
        credentials: "include",
        headers: csrfHeaders("POST"),
        body: fd,
      });
    let res = await doFetch();
    if (res.status === 401 && (await silentRefresh())) res = await doFetch();
    if (!res.ok) throw new ApiError(res.status, await errorMessageFromResponse(res));
    return res.json();
  },
  listProducts: (params?: { vendor_id?: string; active_only?: boolean; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.active_only !== undefined) q.set("active_only", String(params.active_only));
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    if (params?.offset !== undefined) q.set("offset", String(params.offset));
    return get<MealProductOut[]>(`/meal/products?${q}`);
  },
  createProduct: (body: {
    vendor_id: string; name: string; description?: string | null; category?: string | null;
    image_url?: string | null; price: number; default_max_quantity?: number | null;
  }) => post<MealProductOut>("/meal/products", body),
  updateProduct: (id: string, body: Partial<{
    name: string; description: string | null; category: string | null; image_url: string | null;
    price: number; default_max_quantity: number | null; is_active: boolean;
  }>) => patch<MealProductOut>(`/meal/products/${id}`, body),
  listAvailabilities: (params?: {
    vendor_id?: string; date_from?: string; date_to?: string; active_only?: boolean; limit?: number; offset?: number;
  }) => {
    const q = new URLSearchParams();
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.date_from) q.set("date_from", params.date_from);
    if (params?.date_to) q.set("date_to", params.date_to);
    if (params?.active_only !== undefined) q.set("active_only", String(params.active_only));
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    if (params?.offset !== undefined) q.set("offset", String(params.offset));
    return get<MealAvailabilityOut[]>(`/meal/availabilities?${q}`);
  },
  createAvailability: (body: {
    product_id: string; service_date: string; sale_start?: string | null; sale_end?: string | null;
    price?: number | null; max_quantity?: number | null; note?: string | null;
    pickup_slots?: {
      label: string; sort_order?: number; pickup_start: string; pickup_end: string;
      order_deadline: string; capacity?: number | null;
    }[];
  }) => post<MealAvailabilityOut>("/meal/availabilities", body),
  bulkCreateWeeklyAvailabilities: (body: {
    product_ids: string[]; date_from: string; date_to: string; weekdays: number[];
    sale_start_time?: string | null; sale_end_time?: string | null;
    pickup_slots?: {
      label: string; sort_order?: number; pickup_start: string; pickup_end: string;
      order_deadline: string; capacity?: number | null;
    }[];
  }) => post<MealAvailabilityOut[]>("/meal/availabilities/weekly", body),

  // 訂單
  createOrder: (body: {
    schedule_id?: string | null; pickup_slot_id?: string | null;
    items: { menu_item_id?: string | null; availability_id?: string | null; quantity: number }[];
    notes?: string;
  }) =>
    post<MealOrderOut>("/meal/orders", body),
  listOrders: (params?: { my_only?: boolean; schedule_id?: string; vendor_id?: string; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.my_only !== undefined) q.set("my_only", String(params.my_only));
    if (params?.schedule_id) q.set("schedule_id", params.schedule_id);
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    if (params?.offset !== undefined) q.set("offset", String(params.offset));
    return get<MealOrderListItem[]>(`/meal/orders?${q}`);
  },
  getOrder: (id: string) => get<MealOrderOut>(`/meal/orders/${id}`),
  createClassOrder: (body: {
    user_id: string;
    order: {
      schedule_id?: string | null; pickup_slot_id?: string | null;
      items: { menu_item_id?: string | null; availability_id?: string | null; quantity: number }[];
      notes?: string | null;
    };
  }) => post<MealOrderOut>("/meal/orders/class", body),
  updateOrder: (id: string, body: {
    schedule_id?: string | null; pickup_slot_id?: string | null;
    items: { menu_item_id?: string | null; availability_id?: string | null; quantity: number }[];
    notes?: string | null;
  }) => patch<MealOrderOut>(`/meal/orders/${id}`, body),
  cancelOrder: (id: string, reason?: string) =>
    post<MealOrderOut>(`/meal/orders/${id}/cancel`, { reason }),
  confirmOrder: (id: string) => post<MealOrderOut>(`/meal/orders/${id}/confirm`),
  completeOrder: (id: string) => post<MealOrderOut>(`/meal/orders/${id}/complete`),
  setOrderPaid: (id: string, isPaid: boolean) =>
    post<MealOrderOut>(`/meal/orders/${id}/payment?is_paid=${String(isPaid)}`),
  listClassOrders: (params?: { vendor_id?: string; pickup_slot_id?: string; is_paid?: boolean; assisted_only?: boolean; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.pickup_slot_id) q.set("pickup_slot_id", params.pickup_slot_id);
    if (params?.is_paid !== undefined) q.set("is_paid", String(params.is_paid));
    if (params?.assisted_only !== undefined) q.set("assisted_only", String(params.assisted_only));
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    if (params?.offset !== undefined) q.set("offset", String(params.offset));
    return get<MealOrderListItem[]>(`/meal/orders/class?${q}`);
  },
  getClassPickupCode: (params: { class_id: string; vendor_id: string; pickup_slot_id: string }) => {
    const q = new URLSearchParams(params);
    return post<MealClassPickupCodeOut>(`/meal/orders/class-pickup-code?${q}`);
  },
  lookupByCode: (code: string) => get<MealOrderOut>(`/meal/orders/lookup?code=${encodeURIComponent(code)}`),
  pickupLookup: (code: string, redeem = true) =>
    post<MealPickupLookupOut>(`/meal/pickup/lookup?code=${encodeURIComponent(code)}&redeem=${String(redeem)}`),
  getScheduleItemStats: (scheduleId: string) => get<ItemStatOut[]>(`/meal/schedules/${scheduleId}/item-stats`),
  getPickupList: (scheduleId: string) => get<PickupListItemOut[]>(`/meal/schedules/${scheduleId}/pickup-list`),
  assignVendorManager: (vendorId: string, email: string) =>
    post<VendorManagerOut>(`/meal/vendors/${vendorId}/managers`, { email }),
  downloadReport: (format: "xlsx" | "csv", params?: { vendor_id?: string; schedule_id?: string }) => {
    const q = new URLSearchParams();
    if (params?.vendor_id) q.set("vendor_id", params.vendor_id);
    if (params?.schedule_id) q.set("schedule_id", params.schedule_id);
    const qs = q.toString() ? `?${q}` : "";
    return fetch(`${BASE}/meal/reports/orders.${format}${qs}`, {
      credentials: "include",
    });
  },
};
